pkg/rtc/audiolevel.go

```diff
@@ -7,57 +7,57 @@ import (
 
 const (
 	// number of audio frames for observe window
-	observeFrames    = 25 // webrtc default opus frame size 20ms, 25*20=500ms matching default UpdateInterval
+	observeDuration  = 500 // 500ms matching default UpdateInterval
 	silentAudioLevel = 127
 )
 
 // keeps track of audio level for a participant
 type AudioLevel struct {
 	levelThreshold uint8
 	currentLevel   uint32
-	// min frames to be considered active
-	minActiveFrames uint32
+	// min duration to be considered active
+	minActiveDuration uint32
 
 	// for Observe goroutine use
 	// keeps track of current activity
-	observeLevel uint8
-	activeFrames uint32
-	numFrames    uint32
+	observeLevel     uint8
+	activeDuration   uint32 // ms
+	observedDuration uint32 // ms
 }
 
 func NewAudioLevel(activeLevel uint8, minPercentile uint8) *AudioLevel {
 	l := &AudioLevel{
-		levelThreshold:  activeLevel,
-		minActiveFrames: uint32(minPercentile) * observeFrames / 100,
-		currentLevel:    silentAudioLevel,
-		observeLevel:    silentAudioLevel,
+		levelThreshold:    activeLevel,
+		minActiveDuration: uint32(minPercentile) * observeDuration / 100,
+		currentLevel:      silentAudioLevel,
+		observeLevel:      silentAudioLevel,
 	}
 	return l
 }
 
 // Observes a new frame, must be called from the same thread
-func (l *AudioLevel) Observe(level uint8) {
-	l.numFrames++
+func (l *AudioLevel) Observe(level uint8, durationMs uint32) {
+	l.observedDuration += durationMs
 
 	if level <= l.levelThreshold {
-		l.activeFrames++
+		l.activeDuration += durationMs
 		if l.observeLevel > level {
 			l.observeLevel = level
 		}
 	}
 
-	if l.numFrames >= observeFrames {
+	if l.observedDuration >= observeDuration {
 		// compute and reset
-		if l.activeFrames >= l.minActiveFrames {
-			const invObserveFrames = 1.0 / observeFrames
-			level := uint32(l.observeLevel) - uint32(20*math.Log10(float64(l.activeFrames)*invObserveFrames))
+		if l.activeDuration >= l.minActiveDuration {
```

Comment:
A few things to note
1. Can remove the `inv...` variable and just divide in the equation. It is trying to find percentage of active time and that inv and multiply confuses me heavily :-)
2. With the move to duration based, it would be good to take the actual duration while calculating percentage activity. For example, you could be getting packets continuously for 400 ms and then DTX kicks in and the next packet comes after another 400 ms, the duration will be 800 ms which is way more than the constant observe duration.
3. Continuing on the above point, if we want to measure every 1/2 second, we should ideally be running a goroutine and update periodically. If we were to do this, then we do not have worry about the above example as much.

Reply:
Make sense, but do we need to be so precise  here? We can observe and improve it if some people complain with the speech indicator.

Reply:
Yeah, don't need to be precise. We can always wait and watch. But, it is simple enough to make this change now rather than wait. So, thought of suggesting it.